### PR TITLE
process all courses' assignments; misc. other cleaning up (iss. #6, #7, #8, #10)

### LINKED ISSUE #6: expand processing to whole course, multiple courses
The application as developed thus far gets data from a specific assignment in a specific course.  It seems like the next logical step would be to expand the processing to all assignments in a specified course.  It would examine every assignment in the course for peer reviews.

Once support for all assignments in a course is implemented, it's fairly trivial to expand that support to multiple courses, specified in a list.

Rather than checking every rubric of every assignment in each course, it may be faster to check every rubric first.  When any rubrics appears to be for a peer reviewed assignment, then get that assignment.

### LINKED ISSUE #7: update `requirements.txt`
Update `requirements.txt` to…
* Add a version for `canvasapi`
* Use most recent versions of all other modules, especially `Django`

### LINKED ISSUE #8: make debugging code official
Useful debugging code has been written for this project.  Commit it and make it an official part of the project.

### LINKED ISSUE #10: clean up: configuration variables & code format
Fix processing of configuration variables.  E.g., rename `CANVAS_API_URL` → `CANVAS_BASE_URL` and remove UMich `canvas-test` default value.

### PR DESCRIPTION
**Pull request by lsloan**
Resolves #6
Resolves #7
Resolves #8
Resolves #10

## Test Plan
1. Follow the steps in `README.md`.
2. When configuring the `.env` file, set it up with Canvas-test and use the following for course IDs:

   * `COURSE_IDS_CSV=545451`
      Uses "[ECON 101 200 FA 2022](https://canvas-test.it.umich.edu/courses/545451)", which contains three peer-reviewed assignments.
   * `COURSE_IDS_CSV=200,545451,461474`
      Add other courses, which may or may not include peer-reviewed assignments.

### CODE REVIEW

**Comment by lsloan** [comment]
Implementing #6 is more complicated than I expected.  Trying to save peer reviews for some of the other assignments in the test course is revealing more abnormalities in the Canvas data than expected.

File: Dockerfile
Comment (ssciolla) [inline Dockerfile:16]: If this was set up to run as a job using Jenkins/OpenShift, I think you'd need this, otherwise the source code isn't in the image (or container). It shouldn't be a problem to copy this over, I think most of our applications copy and then also mount the directory for local development.

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:32]: Why are we including the test student?

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:50]: I guess this is why the test student is now included. Hm...

**Review by ssciolla — COMMENTED** [review]
Started looking at this. I was able to run it successfully. Still have to review more files and look at the data.

File: Dockerfile
Comment (lsloan) [reply Dockerfile:16]: Thanks for mentioning that.  I'd forgotten about that.  I'll restore the lines.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:32]: Some assessments were made by the test student.  If the test student's ID wasn't included in the user table, then saving its assessments would cause an error about the missing user.

File: Dockerfile
Comment (lsloan) [reply Dockerfile]: Restored!

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:117]: Not sure of all the scenarios here, but this seems like a classic LBYL vs. EAFP scenario. Not sure which approach to go with each scenario, but these seem to be happening with some regularlity, so I'd probably think about using more LBYL. https://realpython.com/python-lbyl-vs-eafp/#eafp-vs-lbyl-with-examples

**Review by ssciolla — COMMENTED** [review]
Still looking. Some things that came to mind.

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:140]: You might consider doing the string splitting and stripping earlier, maybe just do it it `config`.

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:12]: I think you want to just use `getLogger` from `logging` here. I think you're importing it through `canvasData`.

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:153]: Could see what follows becoming a separate function, class/method.

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:193]: Are you waiting to do this because you only want to save the course if it has peer reviews? This seems fine, but would be nice to move it out of the for loop. Maybe you could somehow check whether any of the assignments have peer reviews first, then start looping? But hard to tell in advance if there's actually data without digging into it all. This is probably fine.

**Review by ssciolla — COMMENTED** [review]
Couple more comments. Next I'm looking at the data. I'm seeing a number of "Error saving Assessment" warnings in the logs, basically a missing user or submission. I'll have to see if that's just the nature of the data or something's not quite right.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:140]: True.  It'd be neater if I do it in `config`.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:12]: It is.  I guess it's better not to depend on the imports of the other module.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:153]: Yeah, maybe.  It did grow over time as I was writing it.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:193]: Yeah, I didn't want to save the course and the user data unless we know for sure whether the course's many assignments are peer reviewed and have assessments (peer review comments).  I'll look at it again to see whether it could be done more gracefully.

**Review by ssciolla — COMMENTED** [review]
Was looking at the code locally and figured I'd fix a couple minor things to reduce error noise. Please feel free to use if you'd like (https://github.com/tl-its-umich-edu/peer-review-data/tree/issue-6-type-import-fixes) and delete the branch afterwards.

File: peer_review_data/models.py
Comment (ssciolla) [reply peer_review_data/models.py:1]: After digging into the import situation a bit, it looks like you are defining `db_table` in many cases (not all) by using the internal name of the class from `canvasapi`. It's probably okay, but I guess I'd lean towards hard-coding the names, or letting Django generate them based on the class name, to avoid any possible downstream effects of changes in the library classes or structure.

File: canvasData.py
Comment (zqian) [reply canvasData.py:1]: Eventually those API calls will all be replaced/simplified with the Canvas Data 2 queries, once UDP provides the service towards the end of this year.

File: config/.env.sample
Comment (zqian) [inline config/.env.sample:7]: is '' considered as the same as the the default value 's'?

File: peer_review_data/main.py
Comment (zqian) [inline peer_review_data/main.py:179]: Should the error message be more specific to include the word 'rubric' inside, as 'No peer reviews rubrics assessments were found'?

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:32]: Opened #12.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:50]: Will resolve later, in #12/

File: peer_review_data/models.py
Comment (lsloan) [reply peer_review_data/models.py:1]: If I let Django name the DB tables, it prefixes each one with the name of the app.  For example, the table for the `Assessment` class gets named `peer_review_data_assessment`, which I don't like.

I don't understand why 3/8 of the programmatic table namings didn't work.  They required me to hard-code them.

I'm trying an alternate technique that should give better table names programmatically without the occasional unexplained error.

Comment (ssciolla) [reply peer_review_data/models.py:1]: @lsloan, I'm fairly certain they don't work because of what I said before, when you use `Course.__name__.lower()`, you are referencing `canvasapi.course.Course`, not the class you are currently defining. There's some confusion here because of the star import (`from canvasData import *`). The three ones with errors are classes without equivalents in the `canvasapi` (you subclass object in `canvasData`). Some of this confusion could be avoided I think with more explicit imports. Again, I also don't think it's a great idea to rely on names from `canvasapi` because of the possibility they could change (I think I prefer the hardcoding).

 Perhaps there's a way to reference the class name of the model directly in `Meta`, but I'm not sure what it is.

Comment (lsloan) [reply peer_review_data/models.py:1]: I'm going to leave it as is for now.  I'd like to use a different base class that can programatically generate the table name, but it isn't working right.  I think it's close, but it creates some conflicts.

I'll open a new issue to resolve this.

Comment (lsloan) [reply peer_review_data/models.py:1]: Opened #13.

File: canvasData.py
Comment (lsloan) [reply canvasData.py:1]: Opened #14.

File: peer_review_data/models.py
Comment (lsloan) [reply peer_review_data/models.py:141]: I think it does have the newline at EOF.  I think that change was removing one of **_two_** newlines there.

File: peer_review_data/migrations/0001_initial.py
Comment (lsloan) [reply peer_review_data/migrations/0001_initial.py:1]: No, I was holding off on committing migrations until I was certain the model classes were good.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:117]: Thanks for the suggestion.  I had thought about checking data before trying to save it, but that would be one or more extra steps.  The Django code that saves the model will do its checks even if this code checks ahead of time.  So it seems to me that the less code written to check, the less code that could have bugs.

I agree that I should use more specific exceptions than `Exception`.  I'll look into that and perhaps with parsing the exception messages for data that explain the problem, if the exceptions don't already contain the data in some other field.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: Done.  Resolved in 2edf11d.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:12]: I had been using a logger in `canvasData`, but not any longer.  I didn't notice until I tried to add the import to `main`.  My IDE kept telling me it was unnecessary because of the import through `canvasData`.  I found I didn't use the logger there any more and I also decided I should replace the `import *` with an explicit one.

Comment (lsloan) [reply peer_review_data/main.py:17]: Resolved in 5861955.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:154]: Right.  That is to say, my IDE is more helpful when the variable is explicitly typed.  Completion of class properties wasn't working without it, for example.  It doesn't hurt to have the type declaration there.  In fact, I remember reading some suggestions not long ago that it's helpful stylistically to have it.

File: peer_review_data/models.py
Comment (lsloan) [reply peer_review_data/models.py]: Cool.

I looked at what you have in https://github.com/tl-its-umich-edu/peer-review-data/tree/issue-6-type-import-fixes.  I feel foolish that I didn't notice the various class references in `models.py` are referring to those in `canvasapi`!  That's not what I intended at all.  I meant for each class to refer to **_itself_**.

I've accepted most of the changes you proposed.  As a result, I see now that the naming of the model DB tables was referring to the wrong classes.  Python classes (or their nested classes) cannot refer to themselves, as I intended.  So I've changed the table names to hard-coded strings instead.

File: config/.env.sample
Comment (lsloan) [reply config/.env.sample:7]: The suffix is optional.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:192]: No.  That would be a confusing error message.  The condition isn't looking for rubrics, it's looking for peer reviews.  Or, as Canvas refers to them "assessments".

File: peer_review_data/models.py
Comment (lsloan) [reply peer_review_data/models.py:148]: Yes, I should be more explicit about the Exceptions.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: Resolved in 62af675.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: I've tried to simplify the logic somewhat.  Waiting to save the course until it's known whether there are peer review assignments with assessments still appears to be the correct thing to do.

File: peer_review_data/models.py
Comment (lsloan) [reply peer_review_data/models.py:148]: Resolved in b6c6c30.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: I think this is good as it is.  If anyone thinks this needs to be addressed, let's open a new issue for it.

Comment (lsloan) [reply peer_review_data/main.py]: Replaced `Exception` with more specific exceptions in b6c6c30.

**Comment by lsloan** [comment]
@ssciolla, @zqian: If you get a moment to review this PR again, I'd appreciate it.  It's probably good enough for an approval now, as I've created issues for some of the points that were raised earlier.  I could merge it tomorrow. (I'm away today.)

**Comment by ssciolla** [comment]
@lsloan, hi, sorry, it's been a busy week. I am off tomorrow. I probably won't be able to finish reviewing this until Monday. I'll get to it as soon as I can.

File: peer_review_data/main.py
Comment (ssciolla) [inline peer_review_data/main.py:134]: So as I noted before, these cases are happening pretty frequently. Do you have any ideas why they are happening? User leaves the course for instance? Just curious if you dug into this.

**Review by ssciolla — APPROVED** [review]
Sorry for the delay, @lsloan. This is a pretty meaty pull request, and I hadn't really seen this code before, so I basically felt like I was reviewing the whole project.

I'm approving, with some comments for further thought or future work. It would at least be good to see the GitHub Action fixed or removed before you merge. I don't know with a 100 percent certainty that this is working exactly as intended, but it did collect data, and in the poking around I did, things made sense. Thus I think it's okay to merge and work iteratively on improvements. Looks like some type hints could be tightened up, but I understand it's difficult in some cases since `canvaspi` doesn't seem to have them.

Couple more general suggestions:
1) You may want to add a knob for changing the level of `canvasapi` log messages, since they can be really noisy too like Django's. Something like this should work in `settings.py`:
  ```
    'loggers': {
        'django': {
            'handlers': ['console'],
            'level': os.getenv('LOG_LEVEL', 'INFO').upper(),
            'propagate': False
        },
        'canvasapi': {
            'handlers': ['console'],
            'level': os.getenv('CANVAS_API_LOG_LEVEL', 'WARN').upper()
        }
    }
  ```
2) Part of the difficulty in reviewing this is not knowing how to test the results. To what do I compare the data? What queries ensure that we can do the analysis we need later? I'm not sure what kind of tests this application needs, but selective unit tests and/or the ability to programmatically run certain queries could be useful. Here are some queries I ended up writing to so some of this work (in case they're helpful):
```
-- Courses, assignments, and submissions
select *
from course c
join assignment a
	on a.course_id=c.id
join submission s
	on s.assignment_id=a.id;

-- Assessor gave to submitter these comments about their work in this course
select
	assessor.login_id as `assessor_login_id`,
	submitter.login_id as `submitter_login_id`,
	cm.comments,
	s.id as `submission_id`
from course c
join assignment a
	on a.course_id=c.id
join submission s
	on s.assignment_id=a.id
join `user` submitter
	on s.user_id=submitter.id
join assessment am
	on am.submission_id=s.id
join `user` assessor 
	on am.assessor_id=assessor.id
join comment cm
	on cm.assessment_id=am.id
where c.id='XXXXX';

-- count peer review comments given by each student in a course
select
	assessor.login_id as `assessor_login_id`,
	count(*)
from course c
join assignment a
	on a.course_id=c.id
join submission s
	on s.assignment_id=a.id
join `user` as submitter
	on s.user_id=submitter.id
join assessment am
	on am.submission_id=s.id
join `user` assessor 
	on am.assessor_id=assessor.id
join comment cm
	on cm.assessment_id=am.id
where c.id='XXXXX'
group by assessor.id;
```

3) This process seemed to take about 2 and a half minutes for the large course you suggested we look at. I'm not sure how much we would/will scale this process, but it's possible we reach a point where performance is an issue. Adopting some `async`/`await` or doing bulk database actions could improve the time to complete. Maybe the `async` becomes less relevant if we switch to Canvas Data 2 queries, but something like Django's `bulk_create` would drastically reduce the number of database calls currently made.

**Comment by lsloan** [comment]
@ssciolla, thanks for the approval and further comments.  I'll review them closely tomorrow morning before merging.

**Comment by lsloan** [comment]
@ssciolla, thanks again for the thoughtful, thorough code review.  I really appreciate it.  I understand the delay.

Yes, this PR has a lot going on.  The earlier PRs were mostly about getting the base application going.  Zhen had me pause progress on this project to work on deploying the Longhorn Open peer grading tool.  So when I was able to resume work here, I've been trying to get as much done as possible and quickly before something else comes along to interrupt me.

Addressing the general suggestions in your top-of-review message:

1. Yes, I am dissatisfied with the current logging config.  I think what you've suggested for differing general and `canvasapi` log levels may serve me well for handling #9.  I wasn't sure how to set log levels for just parts of the application.  I'll add this info to that issue.
2. Thanks for the SQL queries.  I had it on my to-do list to make some.  I should've had them in the test plan for this PR and added to the documentation.  I'll be sure to do the latter.  
   This data will be used by the MWrite AI analysis dashboard.  That dashboard already runs using the MPR DB.  This DB has a similar structure, so I think the dashboard's query, maybe somewhat modified, will be what we should use here.  Your queries are similar to that one.
3. I'll make a new issue about improving performance, including a mention of `bulk_create`.

File: config.py
Comment (lsloan) [reply config.py:5]: I noticed in one of your suggestions in the earlier code review, you used an actual data type rather than one from `typing`.  (It might have been `list`.)  I thought that `typing` _should_ be used for this purpose, because it seems to have been created for this.

I'll make a note to review Python documentation about whether using actual types or those from `typing` are recommended.  If you already know of a documented recommendation, please tell me.

File: config.py
Comment (lsloan) [reply config.py:14]: I wasn't sure whether piped types or bracketed ones were preferred now.  I guess I could look at the dates/numbers on the PEPs to know which are the latest.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:114]: Again, I wasn't sure what's the latest recommendation when it comes to specifying types.  I don't like the name of `Optional`, when the value isn't really an option.  I wish they'd named it something like "Nullable" or "Noneable".  

While I was editing that code, I did use the `| None` format you suggested for a little while, but then I thought maybe I should be using `Optional`.  I suppose the piped format is a little shorter and may eliminate the need to import `Optional`, if it's not used anywhere else in the code.

I'll look up recommendations regarding the two.  Again, docs pointers appreciated.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:178]: I thought that if `rubric_settings` was available, it would always have an `id`.  But, I guess strictly speaking, it could be null, as the signature for `get()` says.

File: config.py
Comment (lsloan) [reply config.py:5]: Based on our watercooler conversation, etc., I'll go along with the notion that it's better to use the actual type rather than import something else from `typing`.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:173]: I agree and I wanted to do that, but when I tried it, I found a problem with that.  I'll look at it again and refresh my memory on what happened.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:134]: No, I've not figured out why this happens.  I'd like to know why, but I'm somewhat comfortable with the way the code is handling this.  I've compared the results of this app with those from the Longhorn Open peer grading tool.  In an earlier iteration of my code, the results in the DB were several fewer peer reviews than the number Longhorn found.  I changed some things (don't skip reviews because they didn't weren't the type I thought they should be, etc.) and now this code finds only one less review than Longhorn does.

In a meeting with Instructure people, I mentioned that there were many instances of these missing entities.  They seemed surprised.  I asked whether it could be from the user leaving the course, but they were unsure.  They've asked me to send them example data, so I will do that soon.

File: config.py
Comment (lsloan) [reply config.py]: Resolved in 2f0dc10.

File: config.py
Comment (lsloan) [reply config.py]: Resolved in 2f0dc10.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: Resolved in 2f0dc10.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: Resolved in 2f0dc10.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: It wouldn't hurt, but I don't see much value in it right now.  These entity types go together, so it seemed to me a tiny savings in doing them in the same functions.  If I found some strong reason to separate them, I probably would.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:112]: Yes, I had removed it to test whether data was being missed.  I was trying to find a solution to the number of results between this app and the Longhorn Open app.  Removing the `continue` didn't affect the results at all.  When I saw the log doesn't contain any difference, either, I decided to commit that change.

Now that we're talking about it, though, I can see that there's a possibility that without the `continue`, the app could try to process a non peer-review assessment in ways that don't make sense.  So, I'll put it back.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:73]: No, we don't need to return those.  I'm glad you asked.  In an earlier iteration of the code, I had been returning them, but later I decided it wasn't necessary.  I've removed the `return` statement.

File: .github/workflows/pycodestyle.yaml
Comment (lsloan) [reply .github/workflows/pycodestyle.yaml:25]: Opened #17.

File: peer_review_data/models.py
Comment (lsloan) [reply peer_review_data/models.py:147]: (I wonder why GitHub didn't give me this response box for your question on the conversation page.) ¯\_(ツ)_/¯

This is a leftover from a much earlier iteration of the code.  There had been instances where the assessor or submission objects hadn't been saved or maybe didn't exist at all.  Changing the logic means this is now caught earlier, so I don't need to catch this exception.  I'm removing it.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py]: Resolved in 9c1a794.

File: peer_review_data/main.py
Comment (lsloan) [reply peer_review_data/main.py:73]: Resolved in 9c1a794.

**Pull request merged**
